presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/TwoStatesMapping.java

```diff
@@ -0,0 +1,74 @@
+{
+    private TwoStatesMapping()
+    {
+    }
+
+    private static final Map<Pair<Class<?>, Class<?>>, Class<? extends AccumulatorState>> MAPPING;
+
+    static {
+        MAPPING = new HashMap<>();
+        MAPPING.put(new Pair<>(boolean.class, Block.class), BooleanBlockState.class);
+        MAPPING.put(new Pair<>(boolean.class, boolean.class), BooleanBooleanState.class);
+        MAPPING.put(new Pair<>(boolean.class, double.class), BooleanDoubleState.class);
+        MAPPING.put(new Pair<>(boolean.class, long.class), BooleanLongState.class);
+        MAPPING.put(new Pair<>(boolean.class, Slice.class), BooleanSliceState.class);
+        MAPPING.put(new Pair<>(boolean.class, void.class), BooleanUnknownState.class);
+        MAPPING.put(new Pair<>(double.class, boolean.class), DoubleBooleanState.class);
+        MAPPING.put(new Pair<>(double.class, Block.class), DoubleBlockState.class);
+        MAPPING.put(new Pair<>(double.class, double.class), DoubleDoubleState.class);
+        MAPPING.put(new Pair<>(double.class, long.class), DoubleLongState.class);
+        MAPPING.put(new Pair<>(double.class, Slice.class), DoubleSliceState.class);
+        MAPPING.put(new Pair<>(double.class, void.class), DoubleUnknownState.class);
+        MAPPING.put(new Pair<>(long.class, Block.class), LongBlockState.class);
+        MAPPING.put(new Pair<>(long.class, boolean.class), LongBooleanState.class);
+        MAPPING.put(new Pair<>(long.class, double.class), LongDoubleState.class);
+        MAPPING.put(new Pair<>(long.class, long.class), LongLongState.class);
+        MAPPING.put(new Pair<>(long.class, Slice.class), LongSliceState.class);
+        MAPPING.put(new Pair<>(long.class, void.class), LongUnknownState.class);
+        MAPPING.put(new Pair<>(Slice.class, boolean.class), SliceBooleanState.class);
+        MAPPING.put(new Pair<>(Slice.class, Block.class), SliceBlockState.class);
+        MAPPING.put(new Pair<>(Slice.class, double.class), SliceDoubleState.class);
+        MAPPING.put(new Pair<>(Slice.class, long.class), SliceLongState.class);
+        MAPPING.put(new Pair<>(Slice.class, Slice.class), SliceSliceState.class);
+        MAPPING.put(new Pair<>(Slice.class, void.class), SliceUnknownState.class);
+        MAPPING.put(new Pair<>(void.class, void.class), UnknownUnknownState.class);
+        MAPPING.put(new Pair<>(void.class, Block.class), UnknownBlockState.class);
+        MAPPING.put(new Pair<>(void.class, boolean.class), UnknownBooleanState.class);
+        MAPPING.put(new Pair<>(void.class, double.class), UnknownDoubleState.class);
+        MAPPING.put(new Pair<>(void.class, long.class), UnknownLongState.class);
+        MAPPING.put(new Pair<>(void.class, Slice.class), UnknownSliceState.class);
+    }
+
+    public static Class<? extends AccumulatorState> getStateClass(Class<?> firstState, Class<?> secondState)
+    {
+        Pair<Class<?>, Class<?>> key = new Pair<>(firstState, secondState);
+        if (MAPPING.containsKey(key)) {
```

Comment:
The idiomatic way is to just fetch the value and check for null. This has the benefit of only doing one lookup instead of two. Also, it's safe with Guava Immutable collections since they can't contain nulls.